lib/target/aquarius/classic/aqplusram.asm

```diff
@@ -65,8 +65,11 @@ setup_far_heap:
     push    af                  ;Save it
     srl     a                   ;/4
     srl     a
-    ld      de,+((((CLIB_FARHEAP_FIRST - AQPLUS_FIRST_BANK) >> 2) +1) % 256)
-    ld      hl,0
+    ;
+    ; Bank number is e:hl[15:14], offset is hl[13:0]
+    ;
+    ld      de,CLIB_FARHEAP_FIRST >> 2
```

Comment:
This looks suspicious (if CLIB_FARHEAP_FIRST = 35)

The (virtual) far address space start from `0x010000` and the original intent was trying to map this virtual address onto a physical address, such that it's coherent with bank mappings. 

The original code assumed that the bank corresponding to AQPLUS_FIRST_BANK corresponded to far pointer 0x010000. Thus the shifting was trying to find the far pointer corresponding to CLIB_FARHEAP_FIRST.

Reply:
I left it at 35 since that's what it was originally, there is no reason for it to be 35. I can change it to AQPLUS_FIRST_BANK to match the banking.

Reply:
Actually, the 35 value was chosen based on this:

> Pages 32 through 47 is User RAM and is free for use by BASIC programs.

The assumption being that page 32,33,34 are going to be used by the main program loading, and that we can start banking into 0xc000 from page 35 onwards.

For automatic configuration, it can't be 32 since that will corrupt "main memory" when the far heap is initialised. Given that you've now harmonised the banks, 36 makes more sense since it's the start of a 64k chunk. 

So I think we need to ensure that CLIB_FARHEAP_FIRST is a multiple of 4, then restoring the following should work:

```
    ld      de,+((((CLIB_FARHEAP_FIRST - AQPLUS_FIRST_BANK) >> 2) +1) % 256)
    ld      hl,0
```

So:

* CLIB_FARHEAP_FIRST = 32, then de = 1, so far pointer = 0x010000
* CLIB_FARHEAP_FIRST = 36, then de = 2, so far pointer = 0x020000

Reply:
Yes, this makes sense.

Reply:
The +1 will break the mapping code as pages are no longer numbered 1-n they are 0x20-0x3f. Because of this there is no need to do CLIB_FARHEAP_FIRST - AQPLUS_FIRST_BANK it should be CLIB_FARHEAP_FIRST. So when you shift 0x24 right twice you get 0x90000, so the code below should suffice.

```
ld    de,CLIB_FARHEAP_FIRST >> 2
ld    hl. 0
```

Reply:
Sorry yes, my confusion.

We should still modulus 4 to make life simpler and just add a 64k chunks to the heap apart from any remnant at the end.

Reply:
OK, I think this is good to go.